src/util/mapbox.js

```diff
@@ -100,18 +101,49 @@ export const normalizeTileURL = function(tileURL: string, sourceURL?: ?string, t
     const suffix = browser.devicePixelRatio >= 2 || tileSize === 512 ? '@2x' : '';
     const extension = browser.supportsWebp ? '.webp' : '$1';
     urlObject.path = urlObject.path.replace(imageExtensionRe, `${suffix}${extension}`);
+    urlObject.path = `/v4${urlObject.path}`;
 
-    replaceTempAccessToken(urlObject.params);
-    return formatUrl(urlObject);
+    return makeAPIURL(urlObject);
 };
 
-function replaceTempAccessToken(params: Array<string>) {
-    for (let i = 0; i < params.length; i++) {
-        if (params[i].indexOf('access_token=tk.') === 0) {
-            params[i] = `access_token=${config.ACCESS_TOKEN || ''}`;
+export const canonicalizeTileURL = function(url: string) {
+    const version = "/v4/";
+
+    const urlObject = parseUrl(url);
+    // Make sure that we are dealing with a valid Mapbox tile URL.
+    // Has to be /v4/, with a valid filename + extension
+    if (!urlObject.path.match(version) || !(urlObject.path.match(imageExtensionRe) || urlObject.path.match('.pbf'))) {
+        // Not a proper Mapbox tile URL.
+        return url;
+    }
+    // Reassemble the canonical URL from the parts we've parsed before.
+    let result = "mapbox://tiles/";
+    result +=  urlObject.path.replace(version, '');
+
+    // Append the query string, minus the access token parameter.
+    let ampersand = false;
+    for (const param of urlObject.params) {
+        if (!param.match('access_token')) {
```

Comment:
We should check for string equality here instead of just an arbitrary match, since we only want to exclude the `access_token` parameter, not other parameters that might include the string too.

Reply:
I'm not sure I understand what you mean – the `param` string is both the query param and its value, so checking for string equality wouldn't be possible without splitting up the string. I took Chris' suggestion of making the regex specify the start of the string with an `=`

Reply:
Ah sorry, I thought it was a key/value object